<a href="https://colab.research.google.com/github/DarkAngel285/GoogleCoLab/blob/main/%D0%97%D0%B0%D0%BD%D1%8F%D1%82%D0%B8%D0%B5_21_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Анализ данных. Основы работы
с Pandas. Фильтрация данных



In [2]:
#Подключаем библиотеку
import pandas as pd

In [3]:
#Монтируем гугл диск. Выйдет окно диалога - надо разрешить доступ
from google.colab import drive
drive.mount ('/content/drive')

Mounted at /content/drive


In [4]:
#Здесь будет ваш путь к папке на диске
#Надо залить файл с данными на Google диск, щелкнуть на значок папки слева, найти файл, через контекстное меню скопировать путь и вставить сюда
df = pd.read_csv("/content/drive/MyDrive/GoogleApps.csv")

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Метод info возвращает краткую информацию о каждом столбце и датафрейме в целом.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7352 entries, 0 to 7351
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             7352 non-null   object 
 1   Category        7352 non-null   object 
 2   Rating          7352 non-null   float64
 3   Reviews         7352 non-null   int64  
 4   Size            7352 non-null   float64
 5   Installs        7352 non-null   float64
 6   Type            7352 non-null   object 
 7   Price           7352 non-null   float64
 8   Content Rating  7352 non-null   object 
 9   Last Updated    7352 non-null   object 
 10  Current Ver     7352 non-null   object 
 11  Android Ver     7352 non-null   object 
dtypes: float64(4), int64(1), object(7)
memory usage: 689.4+ KB


Метод head возвращает первые n (по умолчанию 5) строк датафрейма.

In [7]:
print(df.head())

                                                 App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size    Installs  Type  Price Content Rating      Last Updated  \
0      159  19.0     10000.0  Free    0.0       Everyone   January 7, 2018   
1      967  14.0    500000.0  Free    0.0       Everyone  January 15, 2018   
2    87510   8.7   5000000.0  Free    0.0       Everyone    August 1, 2018   
3   215644  25.0  50000000.0  Free    0.0           Teen      June 8, 2018   
4      967   2.8    100000.0  Free    0.0       Everyone     June 20, 2018   

          Current Ver   Android Ver  
0           

Метод tail возвращает последние n (по умолчанию 5) строк датафрейма.

In [8]:
print(df.tail())

                                                App             Category  \
7347                                    Chemin (fr)  BOOKS_AND_REFERENCE   
7348                                  FR Calculator               FAMILY   
7349                               Sya9a Maroc - FR               FAMILY   
7350               Fr. Mike Schmitz Audio Teachings               FAMILY   
7351  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating  Reviews       Size    Installs  Type  Price Content Rating  \
7347     4.8       44   0.604492      1000.0  Free    0.0       Everyone   
7348     4.0        7   2.600000       500.0  Free    0.0       Everyone   
7349     4.5       38  53.000000      5000.0  Free    0.0       Everyone   
7350     5.0        4   3.600000       100.0  Free    0.0       Everyone   
7351     4.5   398307  19.000000  10000000.0  Free    0.0       Everyone   

        Last Updated         Current Ver         Android Ver  
7347  March 23, 2014   

Метод describe возвращает обобщённые статистические показатели на основе хранящихся в датафрейме данных.

In [9]:
df.describe()

Rating       Reviews         Size      Installs        Price
count  7352.000000  7.352000e+03  7352.000000  7.352000e+03  7352.000000
mean      4.177775  3.034096e+05    22.769578  8.662313e+06     1.171476
std       0.544145  1.907619e+06    23.379472  5.132294e+07    17.835302
min       1.000000  1.000000e+00     0.008301  1.000000e+00     0.000000
25%       4.000000  1.017500e+02     5.100000  1.000000e+04     0.000000
50%       4.300000  2.212500e+03    14.000000  1.000000e+05     0.000000
75%       4.500000  3.860625e+04    33.000000  1.000000e+06     0.000000
max       5.000000  4.489389e+07   100.000000  1.000000e+09   400.000000

БОЛЬШОЕ ЗАДАНИЕ для самостоятельной работы.

Как называется приложение, расположенное первым в наборе данных?
Выведи на экран первые несколько строк датасета и определи, как называется приложение (App), данные о котором расположены в первой строке.

In [10]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size    Installs  Type  Price Content Rating      Last Updated  \
0      159  19.0     10000.0  Free    0.0       Everyone   January 7, 2018   
1      967  14.0    500000.0  Free    0.0       Everyone  January 15, 2018   
2    87510   8.7   5000000.0  Free    0.0       Everyone    August 1, 2018   
3   215644  25.0  50000000.0  Free    0.0           Teen      June 8, 2018   
4      967   2.8    100000.0  Free    0.0       Everyone     June 20, 2018   

          Current Ver   Android Ver  
0               1.0.0  4.0.3 and up  
1               2.0.0  4.0.3 and up  
2               1.2.4  4.0.3 and up  
3  Varies with device    4.2 and up  
4                 1.1    4.4 and up

*ОТВЕТ*: Photo Editor & Candy Camera & Grid & ScrapBook

К какой категории относится приложение, расположенное последним в наборе данных?
Выведи на экран последние несколько строк датасета и определи, к какой категории (Category) относится приложение, расположенное в последней строке.

In [11]:
df.tail()

App             Category  \
7347                                    Chemin (fr)  BOOKS_AND_REFERENCE   
7348                                  FR Calculator               FAMILY   
7349                               Sya9a Maroc - FR               FAMILY   
7350               Fr. Mike Schmitz Audio Teachings               FAMILY   
7351  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating  Reviews       Size    Installs  Type  Price Content Rating  \
7347     4.8       44   0.604492      1000.0  Free    0.0       Everyone   
7348     4.0        7   2.600000       500.0  Free    0.0       Everyone   
7349     4.5       38  53.000000      5000.0  Free    0.0       Everyone   
7350     5.0        4   3.600000       100.0  Free    0.0       Everyone   
7351     4.5   398307  19.000000  10000000.0  Free    0.0       Everyone   

        Last Updated         Current Ver         Android Ver  
7347  March 23, 2014                 0.8          2.2 and up  
7348   June 18, 2017               1.0.0          4.1 and up  
7349   July 25, 2017                1.48          4.1 and up  
7350    July 6, 2018                 1.0          4.1 and up  
7351   July 25, 2018  Varies with device  Varies with device

ОТВЕТ: LIFESTYLE


Проанализируй общую информацию о датасете
Выведи на экран общую информацию о датасете и найди ответы на вопросы:
1) Сколько столбцов содержится в наборе данных?
2) Данные какого типа хранятся в каждом из столбцов?

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7352 entries, 0 to 7351
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             7352 non-null   object 
 1   Category        7352 non-null   object 
 2   Rating          7352 non-null   float64
 3   Reviews         7352 non-null   int64  
 4   Size            7352 non-null   float64
 5   Installs        7352 non-null   float64
 6   Type            7352 non-null   object 
 7   Price           7352 non-null   float64
 8   Content Rating  7352 non-null   object 
 9   Last Updated    7352 non-null   object 
 10  Current Ver     7352 non-null   object 
 11  Android Ver     7352 non-null   object 
dtypes: float64(4), int64(1), object(7)
memory usage: 689.4+ KB


ОТВЕТ: 1) 7352; 2) float64(4), int64(1), object(7)

Проанализируй статистические показатели о данных
Выведи на экран результаты вычисления показателей описательной статистики и найди ответы на вопросы:
1) Чему равно среднее арифметическое и медиана размера приложений (Size)?
2) Сколько стоит самое дорогое приложение?
3) *Чему равно среднее арифметическое и медиана количества установок приложений (Installs)?

In [13]:
print(df['Size'].mean(), df['Size'].median(), df['Price'].max(), df['Installs'].mean(), df['Installs'].median(), sep='\n')

22.769577806740514
14.0
400.0
8662312.756800871
100000.0


ОТВЕТ: 1) 22.769577806740514, 14; 2) 400; 3) 8662312.756800871, 100000

Фильтрация данных

<имя_DataFrame>[<имя_столбца>].min() - возвращает минимальное значение признака

In [14]:
print(df['Installs'].min())

1.0


<имя_DataFrame>[<имя_столбца>].max() - возвращает максимальное значение признака

In [15]:
print(df['Installs'].max())

1000000000.0


<имя_DataFrame>[<имя_столбца>].mean() - возвращает среднее значение признака

In [16]:
print(df['Installs'].mean())

8662312.756800871


<имя_DataFrame>[<имя_столбца>].median() - возвращает медианное значение признака

In [17]:
print(df['Installs'].median())

100000.0


In [18]:
avg_price = round(df['Price'].mean(),3)
print('Средняя цена приложения:', avg_price)

Средняя цена приложения: 1.171


Определить среднее количество скачиваний (Installs) приложений, рейтинг (Rating) которых превышает значение 4.9.

In [19]:
print(df[df['Rating'] > 4.9]['Installs'].mean())

331.40697674418607


Определить среднее количество скачиваний (Installs) бесплатных приложений (Type == 'Free'), рейтинг (Rating) которых превышает 4.9.

In [20]:
print(df[(df['Rating'] > 4.9) &(df['Type'] =='Free')]['Installs'].mean())

362.5151515151515


БОЛЬШОЕ ЗАДАНИЕ для самостоятельной работы.

Сколько стоит самое дешёвое платное приложение?
С помощью фильтрации выбери из датасета только платные приложения и найди цену (Price) самого дешёвого из них.

In [21]:
print(df[df['Type'] == 'Paid']['Price'].min())

0.99


ОТВЕТ: 0.99

Чему равна медиана количества установок приложений из категории ART_AND_DESIGN?
С помощью фильтрации выбери из датасета приложения, относящиеся к категории (Category) ART_AND_DESIGN. Найди медиану (median) количества установок (Installs) среди всех отобранных приложений.

In [22]:
print(df[df['Category'] == 'ART_AND_DESIGN']['Installs'].median())

100000.0


ОТВЕТ: 100000

На сколько максимальное количество отзывов для бесплатных приложений больше максимального количества отзывов для платных приложений?
Найди максимальное количество отзывов (Reviews) для бесплатных (Type == 'Free') и платных (Type == 'Paid') приложений. На сколько одно значение больше другого?

In [23]:
print(df[df['Type'] == 'Free']['Reviews'].max() - df[df['Type'] == 'Paid']['Reviews'].max())

44703802


ОТВЕТ: 44703802


Каков минимальный размер приложения для тинейджеров?
С помощью фильтрации выбери из набора данных приложения для тинейджеров (Content Rating == 'Teen').
Каков минимальный размер (Size) приложения в данном наборе?

In [24]:
print(df[df['Content Rating'] == 'Teen']['Size'].min())

0.3154296875


ОТВЕТ: 0.3154296875


*К какой категории относится приложение с самым большим количеством отзывов?
С помощью фильтрации выбери из набора данных приложение с самым большим количеством отзывов (Reviews). Обрати внимание, что в условии для фильтрации ты можешь использовать для этого метод max().
К какой категории (Category) относится найденное тобой приложение?

In [25]:
print(df[df['Reviews'] == df['Reviews'].max()]['Category'].item())

GAME


ОТВЕТ: GAME


*Каков средний рейтинг приложений стоимостью более 20$ и с количеством установок более 10000? С помощью фильтрации по двум условиям выбери из набора данных приложения стоимостью более 20 долларов и с количеством установок более 10000. Определи средний рейтинг таких приложений.

In [26]:
print(df[(df['Price'] > 20) & (df['Installs'] > 10000)]['Rating'].mean())

4.25


ОТВЕТ: 4.25

Методы группировки данных

Метод value_counts() подсчитывает количество уникальных значений признака в столбце.

Посчитать, во сколько раз количество приложений для аудитории «все пользователи»  превышает количество приложений для пользователей «старше 10 лет» при помощи метода value_counts().

In [27]:
temp = df['Content Rating'].value_counts()
print(temp)
everyone = temp['Everyone']
everyone_10 = temp['Everyone 10+']
print(everyone / everyone_10)

Everyone        6166
Teen             868
Everyone 10+     318
Name: Content Rating, dtype: int64
19.38993710691824


Метод groupby() позволяет сгруппировать данные по одному или нескольким признакам и для сгруппированных данных вычислить необходимую статистику.

Посчитать средний размер приложений для всех целевых аудиторий при помощи метода groupby().

In [28]:
print(df.groupby(by = 'Content Rating')['Size'].mean())

Content Rating
Everyone        20.397063
Everyone 10+    40.733419
Teen            33.041953
Name: Size, dtype: float64


Метод aggregate() (сокращённое название agg()) позволяет применить к сгруппированным данным сразу несколько функций для подсчёта статистики. В качестве параметра метод принимает список имён функций в виде строковых констант.

Подсчитать минимальный и максимальный размер приложений для каждой целевой аудитории при помощи метода agg().

In [29]:
print(df.groupby(by = 'Content Rating')['Size'].agg(['min', 'max']))

                     min    max
Content Rating                 
Everyone        0.008301  100.0
Everyone 10+    0.027344  100.0
Teen            0.315430  100.0


БОЛЬШОЕ ЗАДАНИЕ для самостоятельной работы.


Проанализируй приложения по категориям
Сколько всего приложений с категорией ('Category') 'BUSINESS'?

In [30]:
print(df[df['Category'] == 'BUSINESS']['App'].count())

246


ОТВЕТ: 246

Проанализируй соотношение количества приложений
Чему равно соотношение количества приложений для подростков ('Teen') и для детей старше 10 ('Everyone 10+')?
Ответ запиши с точностью до сотых.

In [33]:
print(round(df[df['Content Rating'] == 'Teen']['App'].count() / df[df['Content Rating'] == 'Everyone 10+']['App'].count(), 2))

2.73


ОТВЕТ: 2.73


Проанализируй средний рейтинг приложений по типу
Найди ответы на вопросы:
1. Чему равен средний рейтинг ('Rating') платных ('Paid') приложений? Ответ запиши с точностью до сотых.
2. На сколько средний рейтинг ('Rating') бесплатных ('Free') приложений меньше среднего рейтинга платных ('Paid')? Ответ запиши с точностью до сотых.

In [34]:
print(round(df[df['Type'] == 'Paid']['Rating'].mean(), 2))
print(round(df[df['Type'] == 'Paid']['Rating'].mean() - df[df['Type'] == 'Free']['Rating'].mean(), 2))

4.25
0.08


ОТВЕТ: 1) 4.25; 2) 0.08


Проанализируй минимальный и максимальный размер приложений
Чему равен минимальный и максимальный размер ('Size') приложений в категории ('Category') 'COMICS'?
Запиши ответы с точностью до сотых.

In [37]:
# tmp = df[df['Category'] == 'COMICS']['Size']
# print(tmp.min(), tmp.max(), sep='\n')
print(round(df[df['Category'] == 'COMICS']['Size'].agg(['min', 'max']), 2))

min     0.43
max    40.00
Name: Size, dtype: float64


ОТВЕТ: Минимальный: 0.43; Максимальный: 40


*Бонусные задачи.
Найди ответы на вопросы:
1. Сколько приложений с рейтингом ('Rating') строго больше 4.5 в категории ('Category') 'FINANCE'?
2. Чему равно соотношение бесплатных ('Free') и платных ('Paid') игр с рейтингом ('Rating') больше 4.9?

In [32]:
print(df[(df['Rating'] > 4.5) & (df['Category'] == 'FINANCE')]['App'].count())
print(df[(df['Type'] == 'Free') & (df['Rating'] > 4.9)]['App'].count() / df[(df['Type'] == 'Paid') & (df['Rating'] > 4.9)]['App'].count())

64
8.555555555555555


ОТВЕТ: 1) 64; 2) 8.555555555555555